File: python/iforevents/integrations/mixpanel.py
"""Mixpanel adapter (``pip install "iforevents[mixpanel]"``). Mirrors ``iforevents_mixpanel``."""
from __future__ import annotations

from typing import Any, Optional

from ..events import IdentifyEvent, PageEvent, TrackEvent
from ..integration import Integration
from ._vendor import load_vendor, scalarize


class MixpanelIntegration(Integration):
    """The distinct id is the last identified ``custom_id``; anonymous events use ``anonymous_id``."""

    def __init__(self, token: Optional[str] = None, *, client: Any = None, anonymous_id: str = "server", **hooks: Any) -> None:
        super().__init__(name="MixpanelIntegration", **hooks)
        self.token = token
        self.client = client
        self.anonymous_id = anonymous_id
        self._distinct_id: Optional[str] = None

    def init(self) -> None:
        super().init()
        if self.client is None:
            if not self.token:
                raise ValueError("MixpanelIntegration needs a token or a client")
            self.client = load_vendor("mixpanel", "mixpanel").Mixpanel(self.token)

    def _who(self) -> str:
        return self._distinct_id or self.anonymous_id

    def identify(self, event: IdentifyEvent) -> None:
        super().identify(event)
        self._distinct_id = event.custom_id
        self.client.people_set(event.custom_id, scalarize(event.traits))

    def track(self, event: TrackEvent) -> None:
        super().track(event)
        props = scalarize(event.properties)
        props["time"] = int(event.timestamp.timestamp())
        self.client.track(self._who(), event.name, props)

    def page(self, event: PageEvent) -> None:
        super().page(event)
        props = dict(event.properties, navigation_type=event.navigation_type, to_route=event.to_route, previous_route=event.previous_route)
        self.track(TrackEvent(name=event.name or "page_view", type="page_view", properties=props, timestamp=event.timestamp))

    def reset(self) -> None:
        super().reset()
        self._distinct_id = None

File: ruby/iforevents-mixpanel/lib/iforevents/integrations/mixpanel.rb
# frozen_string_literal: true

require "iforevents"
require "mixpanel-ruby"

module Iforevents
  module Integrations
    # Forwards calls to Mixpanel through mixpanel-ruby. Mirrors iforevents_mixpanel.
    # The distinct id is the last identified custom_id; anonymous events use +anonymous_id+.
    class Mixpanel < Iforevents::Integration
      def initialize(token: nil, tracker: nil, anonymous_id: "server", **hooks)
        super(name: "MixpanelIntegration", **hooks)
        raise ArgumentError, "MixpanelIntegration needs a token or a tracker" if token.nil? && tracker.nil?

        @tracker = tracker || ::Mixpanel::Tracker.new(token)
        @anonymous_id = anonymous_id
        @distinct_id = nil
      end

      def identify(event)
        super
        @distinct_id = event.custom_id
        @tracker.people.set(event.custom_id, scalarize(event.traits))
      end

      def track(event)
        super
        props = scalarize(event.properties)
        props["time"] = event.timestamp.to_i
        @tracker.track(who, event.name, props)
      end

      def page(event)
        super
        props = event.properties.merge("navigation_type" => event.navigation_type, "to_route" => event.to_route, "previous_route" => event.previous_route)
        track(TrackEvent.new(name: event.name, type: "page_view", properties: props, timestamp: event.timestamp))
      end

      def reset
        super
        @distinct_id = nil
      end

      private

      def who
        @distinct_id || @anonymous_id
      end

      def scalarize(hash)
        hash.reject { |_, v| v.nil? }
      end
    end
  end
end
